proposals/0048-container-api.md

```diff
@@ -0,0 +1,196 @@
+- Start Date: 2024-04-26
+- Reference Issues: https://github.com/withastro/roadmap/issues/533
+- Implementation PR: 
+
+# Summary
+
+Astro components are tightly coupled to `Astro` (the metaframework). This proposal introduces a possible server-side API for rendering `.astro` files in isolation, outside the full `astro` build pipeline.
+
+# Example
+
+```js
+import { AstroContainer } from 'astro/container';
+import Component from './components/Component.astro';
+
+const container = await AstroContainer.create();
+console.log(await container.renderToString(Component, { props, slots }))
+```
+
+The container can be optionally constructed with some settings that are typically reserved for Astro configuration.
+```js
+const container = await AstroContainer.create({
+    mode: "production",
+    streaming: true,
+    astroConfig: {
+        site: "https://example.com",
+        trailingSlash: false
+    }
+})
+```
+
+The second argument to `renderToString` optionally provides render-specific data that may be exposed through the Astro global, like `props`, `slots`, `request`, and/or `params`.
+```js
+await astro.renderToString(Component, { props, slots, request, params })
+```
+
+# Background & Motivation
+
+Some of our own proposals, as well as many third-party tool integrations, are blocked until we expose a proper server-side rendering API for `.astro` components. Other frameworks have tools like `preact-render-to-string` or `react-dom/server` that make rendering components on the server straight-forward.
+
+We've avoided doing this because... it's very hard! Part of the appeal of `.astro` components is that they have immediate access to a number of powerful APIs, context about your site, and a rich set of framework renderers. In the full `astro` build process, we are able to wire up all this context invisibly. Wiring up the context manually as a third-party is prohibitively complex.
+
+
+# Goals
+
+- Provide a **low-level** API for rendering `.astro` components in isolation
+- Expose a familiar, user-friendly API
+- Surface enough control for full `astro` parity, but abstract away internal APIs
+- Enable third-party tools to consume `.astro` files on the server
+- Enable unit testing of `.astro` component output
+- Possibly unblock `.mdx` compiledContent/html output?
```

Comment:
In my opinion something cannot be a "possibly" a goal in a stage 3 RFC, same thing for the one below. What this possibly unlock can be discussed in other sections imo.

Reply:
I removed the `mdx` thing. Not in scope for now